qemu/tests/check_link_speed_duplex.py

```diff
@@ -0,0 +1,184 @@
+
+    def run_test_windows(tar_speed, tar_duplex):
+        """
+        Start the test on windows guests. Check whether the speed and duplex
+        information of the guest's nic matches the input params.
+
+        param tar_speed: target speed expected
+        param tar_duplex: target duplex expected
+        """
+        session = vm.wait_for_login(timeout=timeout)
+        error_context.context("Check if the driver is installed and verified",
+                              logging.info)
+        driver_name = params.get("driver_name", "netkvm")
+        run_powershell = (params.get("run_powershell", "yes") == "yes")
+        session = utils_test.qemu.windrv_check_running_verifier(
+            session, vm, test, driver_name, timeout)
+        try:
+            check_speed_duplex_windows(tar_speed, tar_duplex)
+            if run_powershell:
+                check_speed_duplex_windows(tar_speed, tar_duplex,
+                                           method='powershell')
+        finally:
+            session.close()
+
+    def check_speed_duplex_windows(tar_speed, tar_duplex, method='traceview'):
+        """
+        Check the speed and duplex with certain method.
+
+        param tar_speed: target speed expected
+        param tar_duplex: target duplex expected
+        param method: the method to check, one of 'traceview' and 'powershell'
+        """
+        if method == 'traceview':
+            speed, duplex = get_speed_duplex_traceview()
+        elif method == 'powershell':
+            speed, duplex = get_speed_duplex_powershell()
+        else:
+            test.error("Method %s not supported", method)
+        if speed != tar_speed or duplex != tar_duplex:
+            test.fail("The speed and duplex is incorrect in %s, "
+                      "with speed=%s, duplex=%s" % (method, speed, duplex))
+
+    def get_speed_duplex_linux():
+        """
+        Get the speed and duplex information on linux guests.
+
+        return: a tuple of (speed, duplex), which speed is measured by mbps,
+                and duplex is one of 'half' and 'full'.
+        """
+        error_context.context("Get speed & duplex info", logging.info)
+        session = vm.wait_for_login(timeout=timeout)
+        mac = vm.get_mac_address(0)
+        ethname = utils_net.get_linux_ifname(session, mac)
+        check_speed_cmd = params["check_speed_cmd"] % ethname
+        status, output = session.cmd_status_output(check_speed_cmd)
+        if status:
+            test.fail("Failed to get speed info,"
+                      "status=%s, ouput=%s" % (status, output))
+        logging.info(output)
+        result = re.findall(r"(?:Speed:\s+(\d+)Mb/s)|(?:Duplex:\s+(\w+))",
+                            output)
+        if len(result) < 2:
+            test.error("Can't get speed or duplex info")
+        speed = int(result[0][0])
+        duplex = result[1][1].lower()
+        return (speed, duplex)
```

Comment:
Need close session at the end? Is that okay to use one session to go through all function? Not really necessary to open new session and close it again and again in each function. You might only close it every time before vm is destroyed.